### ISSUE TITLE
Customizing html template

### ISSUE THREAD

**Issue by pi0** [nuxt/framework#5565]
History: In Nuxt 2, we were supporting the ability to provide a [custom HTML template](https://github.com/nuxt/nuxt.js/blob/dev/packages/vue-app/template/views/app.template.html) used for the renderer. While it was simple, several issues:

- Only could be customized once by end-users (modules cannot hook to extend it and had to use other methods for injection)
- Can be outdated if we wanted to add a new placeholder for the built-in template
- Additional spaces for customization can break hydration behavior
- With Nitro and moving to standalone server builds, we had to _convert_ it into a JS function
- Different behavior for SSR/SPA is not covered
- Runtime level customizations are not possible
- Opaque behavior without ability to inspect by Nuxt

This is the anatomy of the built-in template as a standard HTML structure: 

``html
<!DOCTYPE html>
<html {{ HTML_ATTRS }}>
  <head {{ HEAD_ATTRS }}>
    {{ HEAD }}
  </head>
  <body {{ BODY_ATTRS }}>
    {{ APP }}
  </body>
</html>
```

The main use-case of providing a custom template is extending placeholders to add a global static script or style or attributes. With Nuxt 3, I'm proposing to have a configuration interface in order to extend template:

```js
export default defineNuxtConfig({
  template: {
    htmlAttrs: [],
    headAttrs: [],
    head: [],
    bodyAttrs: [],
    bodyPrepend: [],
    bodyAppend: []
  }
})
```
(note: we shall make it consistent with [`RenderResult.meta`](https://github.com/nuxt/framework/blob/1b6289f60f70dee0f97f6516ea662e91297a0d3c/packages/nuxt/src/core/runtime/nitro/renderer.ts#L22) interface)

This interface is **forward-compatible**, **predictable**, **easily configurable** and **extandable by modules** and always leads to a valid HTML output.

More dynamic customization and runtime behavior can be provided by `nuxt:render` hooks (https://github.com/nuxt/framework/issues/5534). 

There was also an initial discussion to provide the ability completely override the template as an escape hatch. I believe unless it is proven there is a valid use case that cannot be supported by the Object interface, it is probably safer to not expose it. There might be cases we want to compile something that is not HTML (like a Native output?) but this probably deserves better refactor on renderer implementation to expose an API and allow full overriding renderer file, taking into account it is not rendering HTML.

**Comment by ennioVisco** [comment]
While I can see how intuitive it is for people coming from other frameworks, I'm afraid that an `index.html` (and similarly a full-featured template object in the config) would be a too easy way to introduce bad design practices. Moreover, it will have ambiguous relations with the `useHead`/`useMeta` hooks (what happens if the same script is added to both?). 
I'd be for introducing it **only** in response to a precise use case, and instead, prefer milder solutions to specific issues (e.g. a priority index for scripts that require to be loaded as first/last)

**Comment by Rigo-m** [comment]
> While I can see how intuitive it is for people coming from other frameworks, I'm afraid that an `index.html` (and similarly a full-featured template object in the config) would be a too easy way to introduce bad design practices. Moreover, it will have ambiguous relations with the `useHead`/`useMeta` hooks (what happens if the same script is added to both?). I'd be for introducing it **only** in response to a precise use case, and instead, prefer milder solutions to specific issues (e.g. a priority index for scripts that require to be loaded as first/last)

That's fair.
The issue to solve is this: as of right now, including a third party script inside `nuxt.config.ts` looks like this
```js
{
          hid: "gtm",
          children: `(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
          new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
          j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
          'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
          })(window,document,'script','dataLayer','GTM-XXXXXXX');
          `,
          type: "text/javascript",
},
```

Looks awful and cumbersome. Any ideas on how to make it more elegant?

**Comment by pi0** [comment]
I agree about points of verbosity introduced to `nuxt.config`.

It is probably most convenient if there is a module to add 3rd party scripts and we are working on `@nuxt/script` module, replacing `@nuxtjs/gtm` and other community script modules for the most convenience. 

But I also like the idea of having an `app/` file for partial injection 👍🏼 We can introduce `app/head.html` for `headAppend` shortcut. What do you think?

**Comment by Rigo-m** [comment]
IMHO there should be at least a `app/head.html` for `headAppend`, `app/body-prepend.html` and `app/body-append.html`. I know that you have a lot of granularity with script's `async/defer/fetchpriority` attributes, but many libraries still invites the user to put the scripts as first element of the body or as one of the last. Having a body-prepend and a body-append would be a better DX for users, since they would'n't have to ask themselves any question and just place the script in the place suggested by their service of reference.
Makes sense?

**Comment by pi0** [comment]
I'm not saying it doesn't make sense or usecase. Object API as well as modules (local or external) have the full possibility to inject in all possible positions. And this is probably super temporary since with `@nuxt/scripts` we can cover most of 3rd party integrations in most elegant way. I don't want to blowup `app/` directory with files that might have more rare use in the future.

Do you have some common integration examples can benefit from `body-prepend`? As for `body-append`, putting same script in a nuxt plugin would have the same effect.

**Comment by Rigo-m** [comment]
Google GTM's instructions asks you to place their script as first scripts in the head and firsts in the body:
![Screenshot 2022-06-23 at 15 29 51](https://user-images.githubusercontent.com/1218730/175311712-3e0612ab-180b-4599-be1c-a29528c0c3bf.png)

I don't believe it matters that much, but some users might want to follow guidelines as close as possible I think

**Comment by pi0** [comment]
I see. Indeed makes sense for gtm for exact copy-paste BTW I wouldn't personally add no-script to a Nuxt app since without script Nuxt app is probably not working too and doesn't respect Users privacy when scripts are explicitly disabled by the browser. analytics scripts like to take over with the highest priority regardless of performance implications or user preferences 😅 

Okay then regardless I think this all means we should accelerate to deliver `@nuxt/scripts` module to solve current DX issues and provide best practices out of the box. Module can benefit from object api.

**Comment by ennioVisco** [comment]
> Google GTM's instructions asks you to place their script as first scripts in the head and firsts in the body: ![Screenshot 2022-06-23 at 15 29 51](https://user-images.githubusercontent.com/1218730/175311712-3e0612ab-180b-4599-be1c-a29528c0c3bf.png)
> 
> I don't believe it matters that much, but some users might want to follow guidelines as close as possible I think

It does matter, because GTM could be used to track page events and could even have conditions firing on some `<meta>` info.

Yet, I believe we should still be able to exploit that compile-time pre-processing for optimizations. 
I see more coherent to have something like:
```html
<!-- page.vue -->
<script setup>
const gtm = "XXX"
</script>

<template>
  <div>
    <Head>
      <Script type="text/javascript">
      // code using the {{ gtm }} variable
      </Script>
    </Head>

    <h1>{{ title }}</h1>
  </div>
</template>
```
This would be more inline with nuxt's approach to head management, and still could be reduced to the cases where bad companies like Google force you to copy/paste 3rd-party scripts.

**Comment by Rigo-m** [comment]
If the website is fully statically generated, it should be able to be navigated even when there's no javascript going, therefore `noscript` could make sense. Am I wrong?

**Comment by ennioVisco** [comment]
> If the website is fully statically generated, it should be able to be navigated even when there's no javascript going, therefore `noscript` could make sense. Am I wrong?

That's not true, unfortunately. SSG means that the possible routes result in pre-generated HTML pages, but still, the dynamic data and reactivity will be handled by client-side javascript (the pre-generated pages don't have statically all the information, they are just better for SEO and can avoid server-side logic).

**Comment by jvrsolis** [comment]
Just a comment about my use case, I am using Nuxt 3 and ionic to generate an ios app. While most css classes are standard a-zA-Z the moment I added FormKit (which is meant to be nuxt 3 compatible). It would leave a blank screen due to this error

`SyntaxError: Invalid regular expression: range out of order in character class
`
This is probably due to the $ symbol used in classes for FormKit.

All this to say that if I had a way to update the index.html generated to have a charset metatag of utf-8 the problem would go away. As of now I have to manually add it before deploying to ios every single time I generate the nuxt app or remove a really useful form package.

Just hoping this is resolved soon 👍🏽

**Comment by ennioVisco** [comment]
@Rigo-m btw did you look at [this way](https://v3.nuxtjs.org/guide/directory-structure/plugins#vue-plugins) of setting GTM? 

@jvrsolis this is a bug per-se, did you report it? Meta and few other tags should be the first in the <head>

**Comment by BakoviDagi** [comment]
I think this would also be needed to use a Plugin like this https://github.com/hoiheart/vue-universal-modal#install-plugin

**Comment by danielroe** [comment]
@jvrsolis You should be able to set the charset with:

```js
import { defineNuxtConfig } from 'nuxt'

export default defineNuxtConfig({
  app: {
    head: {
      charset: 'utf-8'
    }
  }
})
```

But utf-8 should actually be the default value and if this is not being generated it's a (separate) bug - do raise a new issue.

**Comment by jvrsolis** [comment]
@danielroe 

I still only get a file with the following when running nuxi generate. This file is what appears in safari and in the ios app inspect. No utf-8 meta tag so blank screen


```
<!DOCTYPE html>
<html >

<head >
  <link rel="stylesheet" href="/_nuxt/entry.6467e0f7.css">
</head>

<body >
  <div id="__nuxt"></div><script>window.__NUXT__={serverRendered:false,config:{public:{theme:"facs",themes:"facs,default",apiProtocol:"http",apiDomain:"api.facsware.local",baseUrl:"http:\u002F\u002Flocalhost:3000",appEnv:"local",appName:"FACS",appVersion:"0.0.0",supportEmail:"support@facsware.com"},app:{baseURL:"\u002F",buildAssetsDir:"\u002F_nuxt\u002F",cdnURL:""}}}</script><script type="module" src="/_nuxt/entry-061cc50e.mjs"></script>
</body>

</html>
```


This is my current nuxt config

```
export default defineNuxtConfig(() => {
	return {
		target: 'static',
		ssr: false,
		alias: {
			'@': resolve(__dirname)
		},
		vite: {
			build: {
				target: 'esnext'
			}
		},
		app: {
			head: {
				charset: 'utf-8',
				titleTemplate: '%s - FACS',
				meta: [
					{ name: 'viewport', content: 'viewport-fit=cover, width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no' }
				]
			}
		},
		css: [
			'@ionic/core/css/core.css',
			'@ionic/core/css/normalize.css',
			'@ionic/core/css/structure.css',
			'@ionic/core/css/typography.css',
			'@ionic/core/css/ionic.bundle.css',
			'@/assets/css/index.scss'
		],
		modules: [
			'@nuxtjs/dotenv',
			'@nuxtjs/tailwindcss',
			'@formkit/nuxt',
			'@intlify/nuxt3'
		],
		intlify: {
			localeDir: 'locales',
			vueI18n: {
			  locale: 'en'
			}
		},
		pwa: {
			icon: {},
			meta: {},
			manifest: {
				name: 'FACS'
			},
			workbox: {
				enabled: true,
				skipWaiting: true
			}
		},
		formkit: {
			configFile: './formkit.config.ts',
		},
		postcss: {
			plugins: {
				tailwindcss: {},
				autoprefixer: {},
			},
		},
		plugins: [],
		experimental: {
			reactivityTransform: true,
			viteNode: true,
		},
		publicRuntimeConfig: {
			theme: process.env.NUXT_PUBLIC_THEME || 'default',
			themes: process.env.NUXT_PUBLIC_THEMES || 'default',
			apiProtocol: process.env.NUXT_PUBLIC_API_PROTOCOL || 'http',
			apiDomain: process.env.NUXT_PUBLIC_API_DOMAIN || 'localhost',
			baseUrl: process.env.NUXT_PUBLIC_BASE_URL || 'http://localhost:3000',
			appEnv: process.env.NUXT_PUBLIC_APP_ENV || 'production',
			appName: process.env.NUXT_PUBLIC_APP_NAME || 'vite',
			appVersion: process.env.NUXT_PUBLIC_APP_VERSION || '0.0.0',
			supportEmail: process.env.NUXT_PUBLIC_SUPPORT_EMAIL || ''
		},
		privateRuntimeConfig: {
			apiSecret: process.env.API_SECRET
		}
	}
});
```

**Comment by danielroe** [comment]
You need to set `ssr: true` to render meta tags.

**Comment by Schetnan** [comment]
We have a need to insert something like the following at the top of our HTML document ( above everything else ).  It is used by a corporate proxy that then uses the information inside of it to  render an appropriate template and header/footer.  in our Nuxt2 apps, we put this snippet at the top of the app.html file.  I am wondering how we might do this in nuxt3 without the app.html or something equivalent?

**Comment by vaban-ru** [comment]
Some news about app.html in Nuxt 3?

**Comment by chuckntaylor** [comment]
This solution above by @pierres worked for me. In my case, I wanted a `div` at the end of the `<body>` for teleporting a fullscreen modal to.

```ts
// server/plugins/nitroPlugin.ts
import { defineNitroPlugin } from 'nitropack/runtime/plugin'

export default defineNitroPlugin((nitroApp) => {
  nitroApp.hooks.hook('render:html', (html, { event }) => {
    html.bodyAppend.push(`<div id="modal-root"></div>`)
  })
})
```

It barked a little about not finding the module 'nitropack/runtime/plugin' but it still worked just fine. No need to import in the `nuxt.config.ts`. I'm using `Nuxt 3.0.0-rc.11` with `Nitro 0.5.4`

**Comment by danielroe** [comment]
I'm glad it's working for you, but I would recommend instead using a `<Teleport to="body">` as Nuxt will handle these automatically: https://v3.nuxtjs.org/api/components/teleports/#example-body-teleport.

**Comment by chuckntaylor** [comment]
Well I feel silly. Still getting the hang of Nuxt.

**Comment by danielroe** [comment]
Not a problem at all! Only mentioned to check whether you had another use case for it.

**Comment by yesworld** [comment]
> This solution above by @pierres worked for me. In my case, I wanted a `div` at the end of the `<body>` for teleporting a fullscreen modal to.
> 
> ```ts
> // server/plugins/nitroPlugin.ts
> import { defineNitroPlugin } from 'nitropack/runtime/plugin'
> 
> export default defineNitroPlugin((nitroApp) => {
>   nitroApp.hooks.hook('render:html', (html, { event }) => {
>     html.bodyAppend.push(`<div id="modal-root"></div>`)
>   })
> })
> ```
> 
> It barked a little about not finding the module 'nitropack/runtime/plugin' but it still worked just fine. No need to import in the `nuxt.config.ts`. I'm using `Nuxt 3.0.0-rc.11` with `Nitro 0.5.4`

@danielroe Could you tell me, please... How to use the ```<Teleport to="body">``` in this example?

**Comment by danielroe** [comment]
@yesworld Check out the documentation I linked above. If you have any particular questions, please create a discussion or ask on Discord, as this issue is about something else.

**Comment by m0dch3n** [comment]
I needed in my `<body>`:

```html
    <div id="teleportdiv"></div>
```

Since I have a very big DOM, teleporting directly to `body` causes a complete `recalculate style` of the <body> node. So adding a modal for example could taking more than > 100 ms of rendering phase when the big DOM was present...

Adding in `app.vue` solved to have the div in the body.

```javascript
  <Teleport to="body">
    <div id="teleportdiv"></div>
  </Teleport>
```

But since I'm doing later 

```javascript
  <Teleport to="#teleportdiv">
    <div>Modal...</div>
  </Teleport>
```

And teleporting to a teleported div doesn't work, in my case the solution was the answer from @chuckntaylor 

```javascript
// /server/plugins/teleportdiv.ts
import { defineNitroPlugin } from 'nitropack/runtime/plugin'

export default defineNitroPlugin((nitroApp) => {
  nitroApp.hooks.hook('render:html', (html, { event }) => {
    html.bodyAppend.push(`<div id="teleportdiv"></div>`)
  })
})
```

So the _'teleportdiv'_ always exists and I can teleport to it...

**Comment by vedmant** [comment]
So Nuxt 3 is out, and now I don't see how I can customize my html, body, adn div#__nuxt tags. There is nothing in documentation about this.

**Comment by kissu** [comment]
@vedmant the team is planning to work on something to make it dead simple: https://nuxt.com/docs/community/roadmap#%EF%B8%8F-roadmap
In the meantime, this is still one of the most relevant topic regarding those questions I guess.

What is exactly your use-case? 😄

**Comment by vedmant** [comment]
@kissu In my case I need too add specific attributes to html, body and app root div.

**Comment by vedmant** [comment]
@kissu That will not be ssr rendered.

**Comment by kissu** [comment]
This GitHub issue + [that page](https://nuxt.com/docs/api/advanced/hooks#lifecycle-hooks) are a good starting point then I guess. 👌🏻

**Comment by vedmant** [comment]
@vedmant Yeah I can do some workaround etc, but there is no straight forward documented way of doing so, like it was in Nuxt 2 https://nuxtjs.org/docs/concepts/views#document-apphtml, this is specifically a problem for migration from Nuxt 2.

**Comment by kissu** [comment]
Yes, we are aware of it, hence my initial answer to you.
Sorry for the inconvenience during your migration. 😞

**Comment by jakubmirejovsky** [comment]
This issue will pop up to anyone who is googling Nuxt3 Google Analytics. GA on it's own are very frustrating, so maybe let's introduce **a way for developers** how to quickly get things done with GA and Nuxt3? :)

**Comment by ennioVisco** [comment]
> This issue will pop up to anyone who is googling Nuxt3 Google Analytics. GA on their own are very frustrating, so maybe let's introduce **a way for developers** how to quickly get things done with GA and Nuxt3? :)

What about this? It works perfectly since RC1 I think. https://nuxt.com/docs/guide/directory-structure/plugins#vue-plugins

**Comment by ThomasBerneSV** [comment]
@danielroe I didn't received an answer. Did you see the discussion?

**Comment by kissu** [comment]
@ThomasBerneSV he probably did, no need to ping him on both sides.

**Comment by DAVIDhaker** [comment]
Are issue alive? Need posibility to modify base HTML template.

**Comment by manniL** [comment]
@DAVIDhaker I think the important parts to modify the base HTML template are there by now

**Comment by patrickmonteiro** [comment]
in my app.vue:

<img width="688" alt="Captura de Tela 2023-01-17 às 15 38 27" src="https://user-images.githubusercontent.com/13258255/212983328-7f674017-b339-41aa-b903-3ed843363562.png">
